ssl/t1_lib.c

```diff
@@ -661,7 +661,12 @@ static const uint16_t tls12_sigalgs[] = {
 
     TLSEXT_SIGALG_dsa_sha256,
     TLSEXT_SIGALG_dsa_sha384,
-    TLSEXT_SIGALG_dsa_sha512
+    TLSEXT_SIGALG_dsa_sha512,
+#endif
+#ifndef OPENSSL_NO_GOST
+    TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256,
+    TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512,
+    TLSEXT_SIGALG_gostr34102001_gostr3411,
```

Comment:
I don't off-hand see any code to exclude these from being offered when the GOST engine is not loaded.

Reply:
They aren't offered in that scenario. During construction of the signature_algorithms extension we add the sigalgs to the WPACKET using the `tls12_copy_sigalgs` function (called in TLSv1.3 too in spite of the name). Part of the purpose of this function is to only include sigalgs which are appropriate for the context. This is done by calling the `tls12_sigalg_allowed` function which performs a number of checks - one of which being to confirm that we recognise the algorithms. If the GOST engine is not loaded then this test fails and the sigalgs are not offerred.

Reply:
Does `ssl_set_sig_mask()` need any updates for new PKEY types?

Reply:
I think, for now, loading the GOST engine should be signal enough.
Recall that to use GOST in TLS, *both sides* have to have the alg enabled.

Reply:
Yes, but if don't want to use GOST as a client, the server accepting GOST does not change that fact.  Sure it won't typically be used, but I that still leaves the attack surface exposed.

Reply:
If I'm not mistaken about the facts, I think that this change should be reverted, and enablement of these sigalgs can be done via the CONF API, with the default list containing only a short list of IETF recommended code points.  (Which might mean that at some we also drop DSA, which only "deserves" its spot for legacy reasons. Not sure when might be a good time to do that).

Reply:
Anyone else have thoughts about this? What should be done?

Reply:
I am not sure that the support of GOST will be added to libcrypto itself soon. 
But I think that these algorithms should be indicated when GOST is enabled and not sent otherwise.

Reply:
@beldmit Thanks for your comment, the key question is what the meaning "When GOST is enabled".  I am not convinced that loading of the engine (or potential inclusion of the algorithms in libcrypto) is a sufficiently direct way to signal that "GOST is enabled" **_for use in TLS_**.

My instinct is that the signal to enable GOST in TLS needs to be explicit. What comes to mind would be an explicit setting of the sigalgs (CONF API or direct SSL API calls) to set a non-default sigalg list.  Especially in TLS 1.3, there may not be a separate signal via inclusion of GOST ciphersuites in the cipher list, and signing algorithms are negotiated exclusively via the sigalgs, so I think that some care is required here...

Reply:
For now I think that GOST is enabled for TLS when +GOST is added to the ciphers list (or ciphersuites are specified directly) and algorithms are available (e.g. via engine). Otherwise it is not.

Reply:
> @beldmit For now I think that GOST is enabled for TLS when +GOST is added to the ciphers list (or ciphersuites are specified directly) and algorithms are available (e.g. via engine). Otherwise it is not.

Is that still true in TLS 1.3? The TLS 1.3 cipherlist does not specify the supported public key algorithms.

Reply:
TLS 1.3 does not define any GOST signature/public-key algorithm.

Reply:
The signature_algorithms extension is shared between TLS 1.2 and TLS 1.3.  IANA does not list any code points for any GOST ciphers or GOST signature algorithms at https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml.  And yet we have `0x00,0x81 ` and `0x00,0x83` as well as the private-use `0xff,0x85` and `0xff,0x87` for GOST.  What does our implementation do with the sigalgs extension in TLS 1.2?  Do we disallow GOST?
Also the currently assigned GOST codepoints are listed as `unassigned` in the IANA registry.  Why are we squatting on these codepoints?
```
#define TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256       0xeeee
#define TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512       0xefef
#define TLSEXT_SIGALG_gostr34102001_gostr3411                   0xeded
```

Reply:
[Updated above comment: s/with the sigalgs extension in TLS 1.2/with the sigalgs extension in TLS 1.3/]

Reply:
Could we limit these extensions to make them applicable for TLS 1.2 only? The GOST ciphersuites have the corresponding maximum TLS version limitation.

Reply:
TLS 1.2 only would make the picture less murky, and the client would should (I hope) then not accept the signature algorithm if it offered no ciphers with a GOST public key algorithm.  Matt can comment on whether perchance we already don't support these sigalgs with TLS 1.3.  The remaining question is then whether it is OK to use the code points in question without IANA assignments.

Reply:
> The remaining question is then whether it is OK to use the code points in question without IANA assignments.

It's pretty murky (as is the case for the the "cryptopro bug" TLS extension); I've always been confused why the relevant technology owners had (apparently) "given up" on getting an actual IANA allocation.  On the balance, it's unclear that a desire for IANA purity outweighs our ABI stability policy, though the latter is somewhat weakened for non-default options, I suppose.

That said, we are just weeks away from a "specification required" policy being in place for these things, which ought to facilitate the process.  But is it appropriate for us to request an allocation on the technology owners' behalf, or do we need to wait for them to do so?

Reply:
Dear all!

In 2015 new Russian GOSTs for block ciphers – "Magma" (=old GOST 28147-89) and new "Kuznyechik" ciphers – and their modes of operations were adopted. Because of many problems connected with this, we had a big lag in specifying our (actually working now in Russia) cipher suites for TLS 1.0-1.2 with the old standards. We hope to close this gap really soon.

The specifications for cipher suites for TLS 1.2 with these new ciphers were agreed in Russian community in April 2018, the Russian standard on them must be adopted before October. 

But for TLS 1.3 we also really need an AEAD mode - and it must be adopted in Russia as a standard this year. After that we'll have everything to specify Russian cipher suites in a IETF document. I hope it will be done in 2019.

Reply:
Where does that leave the codepoint assignments?  Can we be sure of getting the same ones?  And why are a couple of the GOST ciphers in the TLS 1.2 ciphersuite table in the "private use range".  I am hesitant to bake even more unofficial values into a new LTS release.  Can you can you get new code points assigned/reserved promptly?

@kaduk How quickly might sigalg code point reservations be turned out and how well fleshed out must the specification be for that to happen.  Ditto for the TLS 1.2 ciphers.

@mattcaswell Do we support the GOST sigalgs in this PR with TLS 1.3 when a GOST engine is loaded?  Or would they need to be added to yet another table?  And with no public-key specific ciphers iwith TLS 1.3, what "signal" would we use to determine that GOST sigalgs are "enabled"?

Reply:
> Do we support the GOST sigalgs in this PR with TLS 1.3 when a GOST engine is loaded? Or would they need to be added to yet another table? And with no public-key specific ciphers iwith TLS 1.3, what "signal" would we use to determine that GOST sigalgs are "enabled"?

The GOST sigalgs will work in TLSv1.3. However they will only be used if:

1) The client offers them, i.e. the client has loaded the GOST engine
2) The server supports them, i.e. the server has loaded the GOST engine
3) The server has been configured with a GOST certificate

In TLSv1.3 the sigalgs are used to guide certificate selection in order that when the server sends the CertificateVerify message it is signed using an algorithm supported by the client. This is entirely independent of the ciphersuites, which can just use the standard TLSv1.3 ones.

Reply:
Enabling new TLS signature algorithms is an undocumented side-effect of loading the engine. Unbound and BIND tend to support GOST when the engine is available, but there is no expectation of enabling GOST support in the DNSPRIV TLS layer. I am not fond of doing this, especially given the lack of code point assignments, published specifications, ... This side-effect of loading the GOST engine would need to be in the CHANGES file, and would need to be communicated to the DNSPRIV implementors.

Asking users to call `SSL_CTX_set1_sigalgs_list()` is not a good option, because we don't have a token like "RECOMMENDED" or similar to enable use of the sigalgs that you should have by default, and hardcoding of point-in-time lists bites the applications in the backside later.

If applications have to start calling `SSL_CTX_set1_sigalgs_list()` we need a more sophisticated interface.

Reply:
> This side-effect of loading the GOST engine would need to be in the CHANGES file, and would need to be communicated to the DNSPRIV implementors.

This isn't a change. All this PR did was fix something that was broken by an earlier PR. The behaviour is like this in 1.1.0:
https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/ssl/t1_lib.c#L714

So, rather I'd see it the other way around. Changing things to not do this by default would be a breaking change.

Reply:
There is a change, because in 1.1.0 (because there was no TLS 1.3) you also needed to enable GOST ciphers in the client cipherlist.  In  1.1.1 with TLS 1.3, there is no cipherlist constraint, because ciphers are separate from sigalgs.

Reply:
As a hack, we could look for GOST ciphers in the TLS &le; 1.2 cipherlist when deciding whether to enable GOST sigalgs in TLS 1.3, but that'd be a bit gross...

Reply:
My suggestion is that whatever ".cnf" file causes the engine to be loaded can contain additional SSL conf directives to enable GOST in TLS 1.3, or else it stays off.
Secondly, I am still concerned about the code-point squatting in a new LTS release. My instinct is that these require IANA registrations ASAP, which means a specification ASAP. How soon will that happen?

Reply:
As we in Russia have got a specification of the new ciphersuites for TLS 1.2, my colleagues Stanislav Smyshlyaev, Evgeny Alekseev, Ekaterina Smyshlyaeva, Dmitry Eremin-Soleynikov and I have started a work on the IETF draft describing these ciphersuites. We hope to get valid IANA registration for the new ciphersuites as soon as IETF process allows.

Reply:
FYI: the IETF draft that @beldmit has mentioned, is the following one: https://datatracker.ietf.org/doc/draft-smyshlyaev-tls12-gost-suites/